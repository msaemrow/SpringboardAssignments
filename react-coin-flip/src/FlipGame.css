.FlipGame-title{
    margin-bottom: 5px;
}

.FlipGame-button{
    background-color: lightblue;
    border: 2px solid blue;
}

.FlipGame-tails-count{
    margin: 0;
}

.FlipGame-heads-count{
    margin: 0;
}