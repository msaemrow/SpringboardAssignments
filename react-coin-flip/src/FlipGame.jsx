import React, { useState } from "react"
import Coin from "./Coin"
import "./FlipGame.css"

function FlipGame(){
    const [coin, setCoin] = useState({
        side: null,
        heads: 0,
        tails: 0
    })

    function headsOrTails(){
        const randNum = Math.floor(Math.random() * 2);
        const side = randNum === 1 ? "heads" : "tails";
        setCoin(prevState => ({
            ...prevState,
            side: side,
            heads: side === "heads" ? prevState.heads + 1 : prevState.heads,
            tails: side === "tails" ? prevState.tails + 1 : prevState.tails
        }))
    }
    return(
        <div className="FlipGame">
            <h1 className="FlipGame-title">Flip-a-Coin</h1>
            <Coin coinSide={coin.side} />
            <button className="FlipGame-button btn" onClick={headsOrTails}>Flip Coin</button>
            <p className="FlipGame-heads-count">Heads: {coin.heads}</p>
            <p className="FlipGame-tails-count">Tails: {coin.tails}</p>
        </div>
    )
}

export default FlipGame;