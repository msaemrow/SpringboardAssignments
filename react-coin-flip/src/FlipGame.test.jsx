import { render, screen, fireEvent } from "@testing-library/react";
import { expect, it, vi } from "vitest";
import FlipGame from "./FlipGame";


it("loads properly without crashing", () => {
    render(<FlipGame />)
})

it("matches snapshot", () => {
    const { asFragment } = render(<FlipGame />);
    expect(asFragment()).toMatchSnapshot();
})

it("updates tails count correctly with button is clicked", () => {
    const random = vi.spyOn(Math, 'random');

    const { container } = render(<FlipGame />);
    const button = screen.getByText("Flip Coin");
    const tailsCountElement = container.querySelector('.FlipGame-tails-count');
    const headsCountElement = container.querySelector('.FlipGame-heads-count');
    
    random.mockReturnValueOnce(0);
    fireEvent.click(button);
    expect(tailsCountElement).toBeInTheDocument();
    expect(tailsCountElement.textContent).toContain("1");
    expect(headsCountElement).toBeInTheDocument();
    expect(headsCountElement.textContent).toContain("0");

    random.mockRestore();
})
it("updates heads count correctly with button is clicked", () => {
    const random = vi.spyOn(Math, 'random');

    const { container } = render(<FlipGame />);
    const button = screen.getByText("Flip Coin");
    const headsCountElement = container.querySelector('.FlipGame-heads-count');
    const tailsCountElement = container.querySelector('.FlipGame-tails-count');
    
    random.mockReturnValueOnce(0.5);
    fireEvent.click(button);
    expect(tailsCountElement).toBeInTheDocument();
    expect(tailsCountElement.textContent).toContain("0");
    expect(headsCountElement).toBeInTheDocument();
    expect(headsCountElement.textContent).toContain("1");

    random.mockRestore();
})
it("updates heads and tail count correctly with button is clicked multiple times", () => {
    const random = vi.spyOn(Math, 'random');

    const { container } = render(<FlipGame />);
    const button = screen.getByText("Flip Coin");
    const headsCountElement = container.querySelector('.FlipGame-heads-count');
    const tailsCountElement = container.querySelector('.FlipGame-tails-count');
    
    random.mockReturnValueOnce(0.5);
    fireEvent.click(button);
    expect(tailsCountElement).toBeInTheDocument();
    expect(tailsCountElement.textContent).toContain("0");
    expect(headsCountElement).toBeInTheDocument();
    expect(headsCountElement.textContent).toContain("1");

    random.mockReturnValueOnce(0.5);
    fireEvent.click(button);
    expect(tailsCountElement).toBeInTheDocument();
    expect(tailsCountElement.textContent).toContain("0");
    expect(headsCountElement).toBeInTheDocument();
    expect(headsCountElement.textContent).toContain("2");

    random.mockReturnValueOnce(0);
    fireEvent.click(button);
    expect(tailsCountElement).toBeInTheDocument();
    expect(tailsCountElement.textContent).toContain("1");
    expect(headsCountElement).toBeInTheDocument();
    expect(headsCountElement.textContent).toContain("2");

    random.mockRestore();
})

