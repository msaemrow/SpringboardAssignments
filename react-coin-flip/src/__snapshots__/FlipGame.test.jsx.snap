// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="FlipGame"
  >
    <h1
      class="FlipGame-title"
    >
      Flip-a-Coin
    </h1>
    <h5>
      Click the button to start the game!
    </h5>
    <button
      class="FlipGame-button btn"
    >
      Flip Coin
    </button>
    <p
      class="FlipGame-heads-count"
    >
      Heads: 0
    </p>
    <p
      class="FlipGame-tails-count"
    >
      Tails: 0
    </p>
  </div>
</DocumentFragment>
`;
